src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -119,7 +155,7 @@ class UserError(Exception):
     pass
 
 
-class JobValueError(UserError):
+class JobValueError(Exception):
```

Comment:
Because UserError is used in the code specifically for errors that cause the hook to delete the job (see other conversation).

JobValueError, on the other hand, is used in cases where what the scheduler is asking is inconsistent with what the cgroup hook is prepared to do. In almost all the cases where I've seen that, it was either a cgroup hook bug, or an inconsistency -- since the cgroup hook evaluates available resources every time an event is processed but the scheduler _only_ gets its information after an exechost_startup, it is possible for both views to get out of sync -- and not something that could be attributed to the user (since users asking for too many resources are supposed to be stopped by the _scheduler_ from landing on nodes that do not have enough resources.

At several sites the derivation of JobValueError from UserError created black hole nodes just because of a few bytes to a few MB rounding errors when the scheduler tried to use "the entire vnode memory" for a vnode while trying to distribute a superchunk across vnodes. All queues were eventually drained of all jobs and jobs deleted, and users had to resubmit some of their jobs (and figure out which ones had been unlucky to land on black hole nodes and which ones had managed to run). 

In those cases, correct behaviour would have been to requeue and not delete the job, allowing them to try to fit on other nodes, or (if they really could not run after repeated attempts) to be held, so that a site admin could fix discrepancies (or cgroup hook bugs) and then release the jobs.

Reply:
FYI, as of 19.1, the scheduler rounds the amount of memory of a node down to the nearest MB.  This was done to try and create larger node buckets.

Reply:
Sadly that will not be enough in some corner cases. We've sene customers where drivers make 32MB disappear in /sys/devices/system/node/nodeX/meminfo between the time MoM starts up and the time the cgroup hook gets to run a job.

Reply:
Also, rounding to _the nearest_ MB might not be such a great idea in some cases, given some other components might not like it if you rounded up and then placed a superchunk "rounding up" what you allocate on one node: that would make the cgroup hook fail, since you would allocate more than what is seen as available.

These changes to take care of that as well, though, since vnode_hidden_mb now defaults to 1. That takes care of any rounder-uppers, since we'll hide 1MB from the server (and thus scheduler).

Reply:
I'm sorry for the confusion.  I did mean that the scheduler rounded down.  You are correct, rounding up would definitely be a problem.  It also wouldn't do what I was trying to accomplish with trying to coalesce nodes into large node buckets.  This will unfortunately not help where you are losing large chunks of memory like the 32MB case you mentioned.  Should we round down further?  Maybe round down to the nearest 32mb or 64mb?  At some point we'll run into issues with customers trying to request the full amount of memory on a node and the scheduler failing to run it since it thinks there isn't enough.

Reply:
No, I don't think we should round down more than 1MB. There is definitely a common danger in having something that is close to a rounded MB and loses 1MB when rounded down due to small jitter (typically 4KB-16KB), so I would definitely "hide" the last MB.

For more severe cases, such as at ASML (very probably due to the GPFS driver), that's why there is now a vnode_hidden_mb tunable that you can set in the config file. ASML has set this to 64 (after noticing that the scheduler was trying to allocate 32MB more memory than the cgroup hook was prepared to honour) and all their troubles disappeared.